text/Issue 8 - Plug-in based architecture.md

```diff
@@ -0,0 +1,265 @@
+  /// `oven-temperature` is the standardized name.
+  pub keys: Vec<String>,
+
+  phantom: PhantomData<T>
+}
+```
+
+
+## Propagating the changes
+
+We modify the definition of channels to use `Feature<T>` instead of
+`ChannelKind`. Otherwise, this definition is unchanged. We similarly modify
+`add_getter`/`add_setter` to use this new definition of `Channel<T>`.
+
+In the implementation of `Adapter`, type `Value` is replaced by `Any`.
+
+In the Taxonomy API, type `Value` is replaced by `JSON`, with a companion `BinaryParts`.
+
+### Implementation
+
+```rust
+pub enum WatchEvent {
+  EnterRange {
+    from: Id<Getter>,
+    json: JSON,
+    binary: BinaryParts
+  },
+  ExitRange {
+    from: Id<Getter>,
+    json: JSON,
+    binary: BinaryParts
+  },
+  // ...
+}
+
+trait API {
+  fn fetch_values(&self, Vec<GetterSelector>, user: User) ->
+    (ResultMap<Id<Getter>, Option<JSON>, Error>, BinaryParts);
+  fn send_values(&self, TargetMap<SetterSelector, JSON>, binary: &BinaryParts,
+    user: User) -> ResultMap<Id<Setter>, (), Error>;
+}
+```
+
+```rust
+trait Adapter {
+  fn fetch_values(&self, set: &[Id<Getter>], user: User) ->
+    ResultMap<Id<Getter>, Option<Box<Any>>, Error>;
+
+  fn send_values(&self, values: HashMap<Id<Setter>, &Any>, user: User) ->
+    ResultMap<Id<Setter>, (), Error>;
+}
+```
+
+
+# Consequences
+
+## Conflicts
+
+Having the same `Message` in several Adapters causes no conflict.
+
+Several instances of `Feature` may share some or all `keys`, possibly with
+different instances of `Message`. This will commonly be the case for features that
+have not been standardized yet but that are implemented by several
+devices/adapters, or that have been standardized recently. This is considered
+normal. The AdapterManager will handle this, by dispatching per key, rather
+than by `Feature`.
+
+If several Adapters use distinct instances of `Message` for the same `keys`, or
+use the `RawBinary`/`RawJSON` `Message`, this may cause a request to the Foxbox to
+be fail due to a parse error or a serialization error with some adapters and
+succeed for others. Similarly, this may cause a request to the Foxbox to return
+messages with completely unrelated serialization formats for several adapters
+implementing the same key.
+
+We consider that this is an acceptable trade-off for decentralization.
+
+## Performance
+
+The `AdapterManager` can be refactored to attempt to group channels with the
+same key + `Message` together. This will let the `AdapterManager` perform a single
+parsing for each value for all such channels, without loss of performance wrt
+the current solution when the channels all the same precise `Message`
+(i.e. not `RawJSON` or `RawBinary`).
+
+Channels using `RawJSON` or `RawBinary` are expected to be slower than those
+using more precise types.
+
+There will be a small cost associated to using `Any` instead of `Value`, but
+this should be extremely small.
+
```

Comment:
The `Any` vs. `Value` definitely doesn't matter.

The `RawJSON` vs. `RawBinary` probably doesn't matter all that much in the short-to mid-term, but it's good to know that we are not painting ourselves in an unoptimizable corner. I'm not planning to implement that optimization any time soon, mind you.